File: leetcode/113.path-sum-ii/cpp/pathsumii.cpp
#include <vector>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    std::vector<std::vector<int>> pathSum(TreeNode *root, int targetSum)
    {
        return traversal(root, targetSum, {}, 0);
    }

private:
    std::vector<std::vector<int>> traversal(TreeNode *root, int targetSum, std::vector<int> path, int sum)
    {
        if (root == nullptr)
            return {};

        sum += root->val;
        path.push_back(root->val);
        if (root->left == nullptr && root->right == nullptr)
        {
            if (targetSum != sum)
                return {};

            return {path};
        }

        std::vector<std::vector<int>> res;
        auto resLeft = traversal(root->left, targetSum, path, sum);
        auto resRight = traversal(root->right, targetSum, path, sum);
        res.insert(res.end(), resLeft.begin(), resLeft.end());
        res.insert(res.end(), resRight.begin(), resRight.end());

        return res;
    }
};

File: leetcode/113.path-sum-ii/go/pathsumii.go
package pathsumii

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) [][]int {
	return traversal(root, targetSum, []int{}, 0)
}

func traversal(root *TreeNode, targetSum int, path []int, sum int) [][]int {
	if root == nil {
		return [][]int{}
	}

	sum += root.Val
	path = append(path, root.Val)
	if root.Left == nil && root.Right == nil {
		if sum == targetSum {
			res := make([]int, len(path))
			copy(res, path)
			return [][]int{res}
		}
		return [][]int{}
	}

	res := [][]int{}
	res = append(res, traversal(root.Left, targetSum, path, sum)...)
	res = append(res, traversal(root.Right, targetSum, path, sum)...)

	return res
}
